src/libexpr/nixexpr.cc

```diff
@@ -348,73 +360,90 @@ void ExprVar::bindVars(const EvalState & es, const StaticEnv & env)
     this->level = withLevel;
 }
 
-void ExprSelect::bindVars(const EvalState & es, const StaticEnv & env)
+void ExprSelect::bindVars(const EvalState & es, const std::shared_ptr<const StaticEnv> & env)
 {
+    if (debuggerHook)
+        staticEnv = env;
```

Comment:
This has a performance cost in "production" mode, especially the `shared_ptr`, I suspect.
Do these ever get released? A plain pointer and `new` may be low hanging fruit here.

It would also be nice to defer the assignment to the debugger code, so we don't have the (small) overhead here. I guess you could look up the root of the ast with `pos` and work from there? It seems a bit tricky to do this without duplicating the `bindVars` logic though.
If you could conditionally throw an exception when the old `bindVars` impl reaches the breakpoint, it seems you could recover all lost performance.

1. bindVars runs as usual
2. evaluator reaches breakpoint
3. breakpoint sets a private mutable field (`isRetrieveStaticEnvMode = true`)
4. breakpoint invokes `bindVars` (again)
5. `bindVars` reaches the breakpoint
6. `ExprBreakpoint::bindVars` sees `isRetrieveStaticEnvMode == true`, copies the `StaticEnv`, throws it
7. evaluator catches it
8. evaluator starts repl

I guess this doesn't work for stepping though, unless you alter the AST to insert more breakpoints. Yikes.

Maybe we can challenge @pennae for a better idea?

Reply:
The shared_ptrs are released when the Expr tree is released.  There isn't a one-to-one correspondence between Envs and Exprs, so with plain pointers you can't just delete when the Expr is deleted.  In production mode the SharedPtrs are not copied so there shouldn't be much cost involved in freeing them - just the check at the site where the StaticEnv is created, and the `if (debuggerHook)` checks.